File: app/src/main/java/com/dialogapp/dialog/db/MicroBlogDb.java
package com.dialogapp.dialog.db;

import androidx.sqlite.db.SupportSQLiteDatabase;
import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;
import android.content.Context;
import androidx.annotation.NonNull;

import com.dialogapp.dialog.model.AccountResponse;
import com.dialogapp.dialog.model.Item;
import com.dialogapp.dialog.model.MicroBlogResponse;

@Database(entities = {AccountResponse.class, Item.class, MicroBlogResponse.class}, version = 1,
        exportSchema = false)
public abstract class MicroBlogDb extends RoomDatabase {
    public abstract AccountDao accountDao();

    public abstract PostsDao postsDao();

    static MicroBlogDb INSTANCE;

    public static MicroBlogDb getInstance(Context context) {
        if (INSTANCE == null) {
            synchronized (MicroBlogDb.class) {
                if (INSTANCE == null) {
                    INSTANCE = buildDatabase(context);
                }
            }
        }
        return INSTANCE;
    }

    private static MicroBlogDb buildDatabase(Context context) {
        return Room.databaseBuilder(context, MicroBlogDb.class, "microblog.db")
                .fallbackToDestructiveMigration()
                .addCallback(new Callback() {
                    @Override
                    public void onOpen(@NonNull SupportSQLiteDatabase db) {
                        super.onOpen(db);

                        new Thread(() -> {
                            MicroBlogDb microBlogDb = getInstance(context);
                            microBlogDb.beginTransaction();
                            try {
                                microBlogDb.postsDao().pruneUserData();
                                microBlogDb.postsDao().prunePosts();
                                microBlogDb.setTransactionSuccessful();
                            } finally {
                                microBlogDb.endTransaction();
                            }
                        }).start();
                    }
                }).build();
    }
}

File: app/src/main/java/com/dialogapp/dialog/db/MicroBlogDb.kt
package com.dialogapp.dialog.db

import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase
import com.dialogapp.dialog.model.EndpointData
import com.dialogapp.dialog.model.FollowingAccount
import com.dialogapp.dialog.model.Post

@Database(entities = [Post::class, EndpointData::class, FollowingAccount::class],
        version = 2, exportSchema = false)
abstract class MicroBlogDb : RoomDatabase() {

    companion object {

        @Volatile
        private var INSTANCE: MicroBlogDb? = null

        fun getInstance(context: Context): MicroBlogDb? {
            return INSTANCE ?: synchronized(this) {
                INSTANCE ?: buildDatabase(context).also { INSTANCE = it }
            }
        }

        private fun buildDatabase(context: Context): MicroBlogDb {
            return Room.databaseBuilder(context, MicroBlogDb::class.java, "microblog.db")
                    .fallbackToDestructiveMigration()
                    .build()
        }
    }

    abstract fun posts(): PostsDao
    abstract fun endpointData(): EndpointDao
    abstract fun followingData(): FollowingDao
}
